libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Teams/Actions/BaseAuthResponseDialog.cs

```diff
@@ -0,0 +1,188 @@
+            {
+                throw new InvalidOperationException($"A valid {nameof(Title)} is required for {this.GetType().Name}.");
+            }
+
+            var tokenResponse = await GetUserTokenAsync(dc, connectionName, cancellationToken).ConfigureAwait(false);
+            if (tokenResponse != null && !string.IsNullOrEmpty(tokenResponse.Token))
+            {
+                // we have the token, so the user is already signed in.
+                // Similar to OAuthInput, just return the token in the property.
+                if (Property != null)
+                {
+                    dc.State.SetValue(Property.GetValue(dc.State), tokenResponse);
+                }
+
+                // End the dialog and return the token response
+                return await dc.EndDialogAsync(tokenResponse, cancellationToken: cancellationToken).ConfigureAwait(false);
+            }
+
+            // There is no token, so the user has not signed in yet.
+            var activity = await CreateOAuthInvokeResponseActivityAsync(dc, title, connectionName, cancellationToken).ConfigureAwait(false);
+            await dc.Context.SendActivityAsync(activity, cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            // Since a token was not retrieved above, end the turn.
+            return Dialog.EndOfTurn;
+        }
+
+        /// <summary>
+        /// Create a type specific Auth Response using the provided <see cref="CardAction"/>.
+        /// </summary>
+        /// <param name="dc">Current <see cref="DialogContext"/>.</param>
+        /// <param name="cardAction"><see cref="CardAction"/> with a valid Sign In link.</param>
+        /// <returns>An InvokeResponse <see cref="Activity"/> specific to this type of auth dialog.</returns>
+        protected abstract Activity CreateOAuthInvokeResponseActivity(DialogContext dc, CardAction cardAction);
+
+        private static Task<TokenResponse> GetUserTokenAsync(DialogContext dc, string connectionName, CancellationToken cancellationToken)
+        {
+            // When the Bot Service Auth flow completes, the action.State will contain a magic code used for verification.
+            string state = null;
+            if (dc.Context.Activity.Value is JObject valueAsObject)
+            {
+                state = valueAsObject.Value<string>("state");
+            }
+
+            var userTokenClient = dc.Context.TurnState.Get<UserTokenClient>();
+            if (userTokenClient != null)
+            {
+                // CreateTokenResponseFromMessageAsync requires an activity with 'magiccode' in the Text property.
+                var activity = MessageFactory.Text(state);
+                return OAuthHelper.CreateTokenResponseFromMessageAsync(userTokenClient, connectionName, activity, cancellationToken);
+            }
+
+            //TODO: deprecate IExtendedUserTokenProvider
+            if (!(dc.Context.Adapter is IExtendedUserTokenProvider tokenProvider))
```

Comment:
Does this need to be resolved before merging?

Reply:
No, since we are using either 1) the UserTokenClient (from cloud adapter) or 2) the current IExtendedUserTokenProvider (which i think will be deprecated once CloudAdapter goes GA)